src/cachemachine/rubinimageinfo.py

```diff
@@ -0,0 +1,346 @@
+    """
+
+    digest: Optional[str]
+    """image digest for a particular image.
+
+    Example: "sha256:419c4b7e14603711b25fa9e0569460a753c4b2449fe275bb5f89743b01794a30"  # noqa: E501
+    """
+
+    reference_obj: reference.Reference
+    """field containing the results of parsing the reference string with the
+    docker_image class.
+    """
+
+    image_type: RubinImageType
+    """Rubin-specific image type
+
+    Example: RubinImageType.WEEKLY
+    """
+
+    semantic_version: semver.VersionInfo
+    """Semantic version constructed from a tag.  Only comparable between
+    RubinImageInfo instances of the same host, repo, name, and image_type.
+    """
+
+    display_name: str
+    """Human-readable display name corresponding to a tag."""
+
+    cycle: Optional[int]
+    """Cycle (used only in T&S SAL builds)."""
+
+    @classmethod
+    def from_reference(
+        cls,
+        ref: str,
+        recommended_tag: Optional[str] = None,
+        digest: Optional[str] = None,
+    ) -> "RubinImageInfo":
+        """Create a RubinImageInfo object from a reference string."""
+        # Delegate basic parsing to the image reference class
+        if not recommended_tag:
+            recommended_tag = DEFAULT_RECOMMENDED_TAG
+        reference_obj = reference.Reference.parse(ref)
+        # An empty tag should be interpreted as 'latest'
+        if not reference_obj["tag"]:
+            reference_obj["tag"] = "latest"
+        tag = reference_obj["tag"]
+        # If we supplied digest separately, glue it in to the reference
+        #  object as well as the RubinImageInfo object
+        if digest and not reference_obj["digest"]:
+            reference_obj["digest"] = digest
+        contingent_name = reference_obj["name"]
+        host, canonical_name = reference_obj.split_docker_domain(
+            contingent_name
+        )
+        repo, name = canonical_name.split("/", 2)
+        # From here down it is Rubin-specific fields
+        image_type = _type_from_tag(tag, recommended_tag)
+        semver = _semver_from_tag(image_type, tag)
+        display_name = _display_name_from_tag(image_type, tag, recommended_tag)
+        cycle = _cycle_from_semver(semver)
+        return cls(
+            reference=ref,
+            reference_obj=reference_obj,
+            host=host,
+            repo=repo,
+            name=name,
+            tag=tag,
+            digest=reference_obj["digest"],
+            image_type=image_type,
+            semantic_version=semver,
+            display_name=display_name,
+            cycle=cycle,
+        )
+
+    def compare(self, other: "RubinImageInfo") -> int:
+        """This is modeled after semver.compare, but raises an exception
+        if the images are not from the same host, repo, and name, or their
+        image_types are incomparable.
+        """
+        if (
+            self.host != other.host
+            or self.repo != other.repo
+            or self.name != other.name
```

Comment:
shouldn't the host repo and name all be the same since that's the only tags we're looking at?

I'm not sure we should be comparing things between types either?  Only within a type?

Reply:
There's nothing stopping anyone from constructing RubinImageInfo objects from different docker repos and then trying to compare them.  I want that to fail.  I should probably write a test case for that.

Reply:
Specifically I am worried about the case I see on the horizon where T&S want to pull some images from their private repo (with the commercial OpenSplice build) and others from a different repo (Google Artifact Registry or Docker Hub).  I don't think we currently support that in cachemachine but I'll bet you ten bucks we end up needing to before operations.